serialise/src/Codec/Serialise.hs

```diff
@@ -142,6 +151,27 @@ deserialiseOrFail bs0 =
         BS.Empty           -> k Nothing      >>= supplyAllInput BS.Empty
     supplyAllInput _ (CBOR.Read.Fail _ _ exn) = return (Left exn)
 
+-- | Deserialise a Haskell value from the external binary representation,
+-- checking that the representation has been consumed fully with no leftovers,
+-- or get back a @'Either' 'DeserialiseFailure' 'ByteString'@.
+--
+-- @since 0.2.4.0
+deserialiseFullyOrFail :: Serialise a => BS.ByteString -> Either (Either CBOR.Read.DeserialiseFailure BS.ByteString) a
```

Comment:
Do you think we really need the trailing data as a result? Why not just include it into the failure?

If someone really needs to decode a prefix, there are lower level interfaces they can use.

Reply:
I personally am totally fine to be more radical and change the existing behaviour, I was just being conservative as per the convention of "most" open-source projects. The trailing data is already available via the more full CBOR.Read API so we can mention that in a changelog.

Reply:
Including the trailing data is nice as then you can generate a custom error message with some extra details such as the length of the junk or a prefix of it.

As you pointed out, I just realised that `Codec.CBOR.Read.deserialiseFromBytes decode` will have nearly the same signature. However using this requires adding a direct dependency on `cborg` which may be surprising/inconvenient for users. The API in `serialise` otherwise provides no access to the leftover input, so it's nice to have this additional method. For example, the corresponding API in `binary` has it ([decodeOrFail](https://hackage.haskell.org/package/binary-0.8.8.0/docs/Data-Binary.html#v:decodeOrFail)) although it is slightly less clean.

So, I could change the type signature (and definition) to be the same as `Codec.CBOR.Read.deserialiseFromBytes decode` if you prefer?

I deleted a previous comment agreeing with the "more radical" option, since as per what I said above, I now think it's good to have an API in `serialise` that exposes this information - and I suppose changing the existing type signature on `deserialiseOrFail` to express this, would not be acceptable as this would break compilation. Another option is to change `DeserialiseFailure` to include a variant for junk leftovers, which would allow us to preserve the type signature on `deserialiseOrFail` whilst changing its behaviour, but this would break people that are pattern-matching on `DeserialiseFailure`.